### PR TITLE
[RLlib] Torch RL Distributed Trainer

### PR DESCRIPTION
**Pull request by avnishn**
This PR has a working example of how we would use the ray train backend executor on a set of streaming examples. 

I then added a class called the torch rl distributed trainer which shows what the RL trainer potentially will look like. 

RL team PTAL cc @kouroshHakha @smorad @sven1977 @gjoliver 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/rl_trainers/torch_rl_distributed_trainer.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+            train_func=train_loop_per_worker,
+            dataset_spec=empty_spec,
+            checkpoint_manager=checkpoint_manager,
+            checkpoint=None,
+            checkpoint_strategy=None,
+        )
+        self.curr_weights = None
+
+    @classmethod
+    def training_func(self, config):
```
Comment (sven1977) [inline rllib/rl_trainers/torch_rl_distributed_trainer.py:67]: self -> cls

File: rllib/rl_trainers/torch_rl_distributed_trainer.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                result["loss"][str(model)] = loss
+                model_gradients = model.update(loss, get_gradients=self.debug_mode)
+
+            # This is how we will return the weights from the trainer to the
+            # workers
+            result["module_weights"] = ray.put(rl_module.get_state())
+            # session.report is kinda like the return statement after training
+            session.report(result)
+
+    def train(self, dataset):
```
Comment (sven1977) [inline rllib/rl_trainers/torch_rl_distributed_trainer.py:107]: Super nice! So this would basically be the new `learn_on_batch`, iteratively callable method for performing a single train step across all train-workers, correct?

File: rllib/rl_trainers/torch_rl_distributed_trainer.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        self.curr_weights = None
+
+    @classmethod
+    def training_func(self, config):
+        if "rl_module_config" not in config:
+            raise ValueError("rl_module_config not in config")
+        rl_module_config = config["rl_module_config"]
+        rl_module_class = config["rl_module_class"]
+        rl_module = rl_module_class(rl_module_config)
+        queue = config["queues"][session.get_local_rank()]
```
Comment (sven1977) [inline rllib/rl_trainers/torch_rl_distributed_trainer.py:73]: Stupid question: Who will take care of splitting up the data that arrives (as chunks) in these queues?

File: rllib/rl_trainers/torch_rl_distributed_trainer.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+        while 1:
+            data = queue.get()
+            out = []
+            result = {}
+            result["loss"] = {}
+            for model in rl_module.models:
+                # this has to call model.__call__ or model.forward
+                # eventually in order to work with ddp
+                out = model.forward_train(data)
```
Comment (sven1977) [inline rllib/rl_trainers/torch_rl_distributed_trainer.py:96]: Cool! Already using the new APIs.

File: scratch.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        train_func=train_loop_per_worker,
+        dataset_spec=empty_spec,
+        checkpoint_manager=checkpoint_manager,
+        checkpoint=None,
+        checkpoint_strategy=None,
+    )
+
+    data_size = config.get("data_size", 1000)
+    batch_size = config.get("batch_size", 32)
+    train_dataset_factory = lambda: LinearDataset(2, 5, size=data_size)
```
Comment (gjoliver) [inline scratch.py:145]: what should a factory look like for RLlib rollout workers??

File: rllib/rl_trainers/torch_rl_distributed_trainer.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        num_models = rl_module.num_models
+
+        pgs = [torch.distributed.new_group(list(range(session.get_world_size()))) for _ in range(num_models)]
+        moved_models = []
+        for pg, model in zip(pgs, rl_module.models):
+            model.to(device)
+            model = DDP(model, device_ids=[session.get_local_rank()], process_group=pg)
+            moved_models.append(model)
+
+        rl_module.set_models(moved_models)
```
Comment (sven1977) [inline rllib/rl_trainers/torch_rl_distributed_trainer.py:85]: Regarding this `set_models()`: Where would the "master copy" of the rl_module sit in the end?

Like right now, Policy keeps a copy on the CPU (+ n copies on the n GPUs). All n GPUs' gradients are averaged and applied to the CPU copy. Before a new loss backprop step, the GPU copies are synched from the CPU one.

File: scratch.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+    from ray.data.datasource import SimpleTorchDatasource
+
+    dataset = ray.data.read_datasource(
+        SimpleTorchDatasource(), parallelism=1, dataset_factory=train_dataset_factory
+    )
+    shards = dataset.repartition(num_blocks=scaling_config.num_workers).split(scaling_config.num_workers)
+
+
+    for i in range(3):
```
Comment (sven1977) [inline scratch.py:157]: `(3)` -> `epochs`

File: rllib/rl_trainers/torch_rl_distributed_trainer.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+        backend_executor = BackendExecutor(
+                backend_config=backend_config,
+                num_workers=scaling_config.num_workers,
+                num_cpus_per_worker=scaling_config.num_cpus_per_worker,
+                num_gpus_per_worker=scaling_config.num_gpus_per_worker,
+                max_retries=0,
+            )
+
+        checkpoint_manager = CheckpointManager(
```
Comment (sven1977) [inline rllib/rl_trainers/torch_rl_distributed_trainer.py:46]: Dumb question: Would this conflict with our own checkpoint handling? RLlib algos have to checkpoint more than just the trainer state.

File: rllib/rl_trainers/torch_rl_distributed_trainer.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+from ray.train._internal.utils import construct_train_func
+from ray.train._internal.backend_executor import BackendExecutor
+from ray.train.torch import TorchConfig
+from ray.train import TrainingIterator
+from ray.train._internal.checkpoint import CheckpointManager
+from ray.train._internal.dataset_spec import RayDatasetSpec
+from ray.util.queue import Queue
+
+
+class TorchRLDataParallelTrainer:
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/rl_trainers/torch_rl_distributed_trainer.py:20]: correct. self.trainer will be accessible in algorithm. It will always have a marl_trainer interface from algorithm's view point.

Comment (kouroshHakha) [inline rllib/rl_trainers/torch_rl_distributed_trainer.py:20]: Some high-level interface stuff to make this consistent with the current design of RLTrainer, This is the abstraction that is expected from an RLTrainer:
```
@abc.abstractmethod
	update(batch: BatchType, **kwargs) -> Any → Performs an update and returns anything that might be needed inside the algorithm.

```

In multi-gpu case you pass in a batch of data and it should automatically  distribute that batch into sub-batches in the Queue object that was created earlier. The trainer actors will update the model as soon as the queue get's field. But who enqueues the data into those queues? trainer.update(batch) should do this.

File: rllib/rl_trainers/torch_rl_distributed_trainer.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        rl_module_class = config["rl_module_class"]
+        rl_module = rl_module_class(rl_module_config)
+        queue = config["queues"][session.get_local_rank()]
+
+        device = train.torch.get_device()
+        num_models = rl_module.num_models
+
+        pgs = [torch.distributed.new_group(list(range(session.get_world_size()))) for _ in range(num_models)]
+        moved_models = []
+        for pg, model in zip(pgs, rl_module.models):
```
Comment (kouroshHakha) [inline rllib/rl_trainers/torch_rl_distributed_trainer.py:80]: Instead of iterating through models inside RLModule let's not put any constraints on how RLModule should contain its submodules. Let's assume RLModule is the simple abstraction below:

```
class RLModule(nn.Module):
    def __init__():
          self.model1 = ...
          self.model2 = ...
    def forward_train(batch):
        # user can specify any arbitrary computational graph between submodules and then return a dict specifying everything that needs to be stored from a single forward pass for model updates
        return dict()
    def forward(batch):
        # this will be an alias to forward_train() since you mentioned DDP's assumption is that the model has a forward() interface.
        return forward_train(batch)
```

With this API you only have to ddpify the parent RLModule instead of it's children which could have implications on how its children are constructed.

File: rllib/rl_trainers/torch_rl_distributed_trainer.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        while 1:
+            data = queue.get()
+            out = []
+            result = {}
+            result["loss"] = {}
+            for model in rl_module.models:
+                # this has to call model.__call__ or model.forward
+                # eventually in order to work with ddp
+                out = model.forward_train(data)
+                loss = model.loss(out)
```
Comment (kouroshHakha) [inline rllib/rl_trainers/torch_rl_distributed_trainer.py:97]: Model's are not aware of their losses. RLTrainer would not about the loss functionality.

File: rllib/rl_trainers/torch_rl_distributed_trainer.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+                result["loss"][str(model)] = loss
+                model_gradients = model.update(loss, get_gradients=self.debug_mode)
```
Comment (kouroshHakha) [inline rllib/rl_trainers/torch_rl_distributed_trainer.py:99]: Same thing here. Model's don't know how to update themselves. RLTrainer does.
